### DISCUSSION TITLE
How to run XUnit tests in parallel with scope?

### DISCUSSION THREAD

**Discussion by A-Rai-col** [xunit/xunit#2924] [Question] [answered]
## Question ❓
 I'm planning to use XUnit for integration tests, where only tests within the same namespace are executed in parallel and the other namespaces sequentially wait for the previous namespace to finish their run, before running their tests in parallel and so forth. 
I was wondering if the following is possible to achieve using XUnit, if so how?

---
## Description: 

Here is how my project will be structured. 

```
TestProject/
├── TopLevelSetup.cs
├── ConfigA/
│   ├── Setup.cs 
│   ├── ATest1.cs
│   ├── ATest2.cs
│   └── ATest3.cs
├── ConfigB/
│   ├── Setup.cs 
│   ├── BTest1.cs
│   ├── BTest2.cs
│   └── BTest3.cs
└── ConfigC/
    ├── Setup.cs 
    ├── CTest1.cs
    ├── CTest2.cs
    └── CTest3.cs
```

* Tests classes are separated at a namespace level e.g. ConfigA, ConfigB, ConfigC. 
* Each namespace will have its own `Setup.cs` class, which the test classes and methods in the same namespace depend on.
* All tests classes/methods need to wait for its namespace level setup to complete before being run. 
* After `Setup.cs` is complete all child test classes and methods need to run in parallel. 
* Each namespace should wait for previous namespace tests to complete before running its tests. (any order)
 
ℹ`TopLevelSetup.cs` is a ClassFixture that will be used to share context between namespaces whereas `Setup.c`s will be used to share context within its own namespace only. 

--- 
## NUnit Example demo 
I've managed to achieve this functionality using NUnit, [Here](https://github.com/A-Rai-col/DiscussionTestProject) is the repo with the NUnit example. 
Each relevant step has a random `thread.sleep` to simulate delay and logs its name to Tests output.

Here are the logs produced when you run all the tests. 
```
NUnit Adapter 4.5.0.0: Test execution started
TopLevelSetup.OTSetup
ConfigA.Setup.OTSetup - Start setup
ConfigA.Setup.OTSetup - Finished Setup
ATest1 Setup
ATest3 Setup
ATest2 Setup
ATest1 Setup
ATest1 Setup
ATest2 Setup
ATest3 Setup
ATest3 Setup
ATest2 Setup
ATest1_Test2
ATest3_Test1
ATest1_Test1
ATest1_Test3
ATest3_Test2
ATest3_Test3
ATest2_Test1
ATest2_Test2
ATest2_Test3
ATest3 TearDown
ATest3 TearDown
ATest2 TearDown
ATest2 TearDown
ATest1 TearDown
ATest2 TearDown
ATest1 TearDown
ATest3 TearDown
ATest1 TearDown
ConfigA.Setup.OTTearDown
ConfigB.Setup.OTSetup - Start setup
ConfigB.Setup.OTSetup - Finished Setup
BTest1 Setup
BTest2 Setup
BTest1_Test1
BTest1 Setup
BTest1 Setup
BTest2_Test2
BTest2 Setup
BTest1_Test3
BTest1_Test2
BTest3 Setup
BTest3_Test2
BTest3 Setup
BTest2_Test1
BTest3 Setup
BTest2 Setup
BTest3_Test1
BTest2_Test3
BTest3_Test3
BTest1 TearDown
BTest3 TearDown
BTest2 TearDown
BTest1 TearDown
BTest3 TearDown
BTest3 TearDown
BTest2 TearDown
BTest1 TearDown
BTest2 TearDown
ConfigB.Setup.OTTearDown
ConfigC.Setup.OTSetup - Start setup
ConfigC.Setup.OTSetup - Finished Setup
CTest2 Setup
CTest3 Setup
CTest3 Setup
CTest1 Setup
CTest2_Test1
CTest2 Setup
CTest1_Test1
CTest3_Test2
CTest3_Test1
CTest3 Setup
CTest2 Setup
CTest2_Test2
CTest1 Setup
CTest3_Test3
CTest2_Test3
CTest1 Setup
CTest1_Test2
CTest1_Test3
CTest1 TearDown
CTest2 TearDown
CTest1 TearDown
CTest2 TearDown
CTest3 TearDown
CTest3 TearDown
CTest1 TearDown
CTest2 TearDown
CTest3 TearDown
ConfigC.Setup.OTTearDown
TopLevelSetup.OTTearDown
NUnit Adapter 4.5.0.0: Test execution complete
```

**Comment by bradwilson** [comment]
Our parallelization right now is built on the concept of "test collections": Tests within the same collection do not run in parallel against each other, but tests from different collections can run in parallel. What you're asking for is basically an inversion of how our parallelism logic currently behaves (you want to group things for parallelism rather than sequentiality). More complicated here is that you're also asking for what is effectively a "collection fixture" (that's the Setup/Teardown you're doing at the group level), which in our model is further indication of sequentiality (because it's intended to share a piece of state across all the tests).

Could it be done via extensibility? Almost surely. Would it be easy? Definitely not easy to figure out for someone who doesn't know our extensibility system well, which means mostly... just me. 😭😂

Let me see how difficult this would be, and if it's easily doable, I'll set it up as a sample project that you could then steal the infrastructure pieces from. It's gonna take a little time.

**Reply by bradwilson** [reply, marked as answer]
I have a sample that I believe does what you want. Attached below is the output I'm seeing. I have applied no sort order to the namespaces, so they're running in the order that they were discovered via reflection (or perhaps the order that LINQ chose to group them in), which is C => B => A. (I commented out all the `Assert.Fail()` calls because they were just adding noise to the output.)

I had to override a bunch of the execution pipeline so that things got flipped from sequential to parallel and parallel to sequential. The sample does not respect any of the fixture interfaces nor test collection definitions, but instead creates a test collection per namespace, and then uses that to ensure that we run the namespaces sequentially. Everything else (Test Class => Test Method => Test Case) becomes parallelized. Although your example didn't show it, I also provided parallel support for individual theory data rows (and one sample class that has that, `ATest4`, so you can see it all running in parallel).

Obviously, I tested this with your happy path (included in the sample) but have not stressed it to ensure that it fails in the right ways when necessary (aka, no unit tests).

Source: https://github.com/xunit/samples.xunit/tree/main/NamespaceParallelization

```
TopLevelSetup Initialize
ConfigC.Setup - Start Setup
ConfigC.Setup - Finish Setup
CTest1 Initialize
CTest3 Initialize
CTest1 Initialize
CTest2 Initialize
CTest2 Initialize
CTest1 Initialize
CTest2 Initialize
CTest3 Initialize
CTest3 Initialize
CTest1_Test1
CTest1_Test3
CTest3_Test1
CTest1_Test2
CTest2_Test1
CTest2_Test3
CTest3_Test3
CTest3_Test2
CTest2_Test2
CTest2 Dispose
CTest3 Dispose
CTest3 Dispose
CTest3 Dispose
CTest2 Dispose
CTest1 Dispose
CTest1 Dispose
CTest2 Dispose
CTest1 Dispose
ConfigC.Setup - Dispose
ConfigB.Setup - Start Setup
ConfigB.Setup - Finish Setup
BTest1 Initialize
BTest2 Initialize
BTest2 Initialize
BTest1_Test3
BTest2_Test2
BTest2_Test3
BTest2 Initialize
BTest2_Test1
BTest1 Initialize
BTest3 Initialize
BTest1 Initialize
BTest1_Test1
BTest3_Test3
BTest1_Test2
BTest3 Initialize
BTest3_Test1
BTest3 Initialize
BTest3_Test2
BTest2 Dispose
BTest1 Dispose
BTest1 Dispose
BTest3 Dispose
BTest2 Dispose
BTest2 Dispose
BTest1 Dispose
BTest3 Dispose
BTest3 Dispose
ConfigB.Setup - Dispose
ConfigA.Setup - Start Setup
ConfigA.Setup - Finish Setup
ATest1 Initialize
ATest3 Initialize
ATest2 Initialize
ATest2 Initialize
ATest1_Test3
ATest3_Test3
ATest2_Test3
ATest2 Initialize
ATest2_Test1
ATest2_Test2
ATest1 Initialize
ATest3 Initialize
ATest1_Test2
ATest3_Test1
ATest1 Initialize
ATest3 Initialize
ATest1_Test1
ATest3_Test2
ATest4 Initialize
ATest4 Initialize
ATest4 Initialize
ATest4 Initialize
ATest4 Initialize
ATest4_Test("Hello, World!") starting
ATest4_Test(42) starting
ATest4_Test(null) starting
ATest4_Test(21.12) starting
ATest4_Test(Object { }) starting
ATest3 Dispose
ATest1 Dispose
ATest4_Test("Hello, World!") finished
ATest2 Dispose
ATest4 Dispose
ATest3 Dispose
ATest2 Dispose
ATest4_Test(null) finished
ATest4 Dispose
ATest3 Dispose
ATest1 Dispose
ATest1 Dispose
ATest2 Dispose
ATest4_Test(42) finished
ATest4 Dispose
ATest4_Test(21.12) finished
ATest4 Dispose
ATest4_Test(Object { }) finished
ATest4 Dispose
ConfigA.Setup - Dispose
TopLevelSetup Dispose
```

**Reply by A-Rai-col** [reply]
Thanks Brad! 🙌
 I appreciate the comments you've added to the extensions as well.

> Could it be done via extensibility? Almost surely. Would it be easy? Definitely not easy to figure out for someone who doesn't know our extensibility system well, which means mostly... just me. 😭😂

Almost everyone mentions how extensible XUnit is but looking at the example you've setup, nope not easy at all 😅.

**Reply by bradwilson** [reply]
This is a category of extension that's basically like saying "I want to make my own test framework". We're defining a whole new set of rules and behaviors about how things are supposed to work, and the extensibility does allow for that, but makes you work for it a little bit. 😄 

Most extensibility is not this level of rule-rewriting.
